lib/router/related-tasks.js

```diff
@@ -0,0 +1,131 @@
+const { isUndefined } = require('lodash');
+const { Router } = require('express');
+const router = Router({ mergeParams: true });
+
+const { BadRequestError, UnauthorisedError } = require('@asl/service/errors');
+const { open } = require('../flow');
+const { autoResolved } = require('../flow/status');
+const filterToEstablishments = require('../query-builders/filter-to-establishments');
+
+module.exports = taskflow => {
+  const Task = taskflow.Task;
+
+  const relatedTasksQuery = ({ user, model, modelId, establishmentId, onlyOpen, sort = { column: 'updatedAt' }, limit, offset }) => {
+    let query = Task.query();
+
+    switch (model) {
+      case 'establishment':
+        // any changes to the establishment licence, including places and roles
+        const id = parseInt(modelId, 10);
+
+        query.whereJsonSupersetOf('data', { establishmentId: id });
+        query.where(builder => {
+          builder.whereJsonSupersetOf('data', { model: 'establishment' });
+          builder.orWhereJsonSupersetOf('data', { model: 'place' });
+          builder.orWhereJsonSupersetOf('data', { model: 'role' });
+        });
+        break;
+
+      case 'profile-touched':
+        // any tasks that have involved the profile (e.g. licence holder, application submitter, profile update, role assignment)
+        // seems fine locally but possibly slow with lots of tasks - might be worth monitoring query time in the logs
+        query.whereRaw('cases.data::text LIKE ?', [`%${modelId}%`]);
```

Comment:
What alternatives to this do we have? I can't image for a second it's going to work nicely against the number of records in prod.